include/envoy/network/listen_socket.h

```diff
@@ -55,13 +55,14 @@ class Socket {
      */
     virtual void hashKey(std::vector<uint8_t>& key) const PURE;
   };
-  typedef std::unique_ptr<Option> OptionPtr;
-  typedef std::shared_ptr<std::vector<OptionPtr>> OptionsSharedPtr;
+  typedef std::shared_ptr<const Option> OptionConstSharedPtr;
```

Comment:
Can you comment on why this needs to be a shared_ptr (not unique)? I skimmed the code but didn't see where/why they need to be shared.

Reply:
+1, why dos this need to be shared?

Reply:
This is now an immutable object which benefits from shared const ownership; we do this already for address instances and some other cases. `unique_ptr` was problematic given how the `OptionPtr` vector is combined in https://github.com/envoyproxy/envoy/pull/2922/files#diff-cd280606785949980237b2a7f08b7885R95, you would need some way to copy/clone the objects across the lists otherwise.

Reply:
I see where you need the shared_ptr in the vector, the reason I didn't like was a shared_ptr of shared_ptr vector `OptionsSharedPtr`, if it is inevitable I'm fine.